File: backend/backend/Service/UserService.cs
using backend.DbContext;
using backend.DTO;
using backend.Model;
using backend.Helpers;
using System;
using System.Collections.Generic;
using System.Linq;

namespace backend.Service
{
    public class UserService
    {
        private readonly AppDbcontext _context;

        public UserService(AppDbcontext context)
        {
            _context = context;
        }

        // Get all users
        public IEnumerable<UserDTO> GetAllUsers()
        {
            return _context.Users.Select(user => new UserDTO
            {
                Id = user.Id,
                Email = user.Email,
                Password = user.Password,
                PasswordSalt = user.PasswordSalt,
                FirstName = user.FirstName,
                LastName = user.LastName,
                TS = user.Ts,
                Active = user.Active
            }).ToList();
        }

        // Get a user by ID
        public UserDTO GetUserById(int id)
        {
            var user = _context.Users.Find(id);
            if (user == null) return null;

            return new UserDTO
            {
                Id = user.Id,
                Email = user.Email,
                Password = user.Password,
                PasswordSalt = user.PasswordSalt,
                FirstName = user.FirstName,
                LastName = user.LastName,
                TS = user.Ts,
                Active = user.Active
            };
        }

        // Create a new user
        public void CreateUser(UserDTO userDto)
        {
            var user = new User
            {
                Email = userDto.Email,
                Password = userDto.Password,
                PasswordSalt = userDto.PasswordSalt,
                FirstName = userDto.FirstName,
                LastName = userDto.LastName,
                Ts = DateTime.Now,
                Active = userDto.Active
            };

            _context.Users.Add(user);
            _context.SaveChanges();
        }

        // Update an existing user
        public bool UpdateUser(int id, UpdateUser userDto)
        {
            var user = _context.Users.Find(id);
            
                if (user == null) return false;
            user.Email = userDto.Email;
            user.FirstName = userDto.FirstName;
            user.LastName = userDto.LastName;
            user.Ts = DateTime.Now;

            _context.Users.Update(user);
            _context.SaveChanges();
            return true;
        }
        public SignUpResponse UpdateCurrentUser(int id, UpdateCurrentUser userDto){
            var user = _context.Users.Find(id);
            if(user ==null) {
                return new SignUpResponse {
                Success = false,
                Error = "No User", 
                ErrorCode = "503"
                };
            }
                if (userDto.Password != userDto.ConfirmPassword)
                {
                    return new SignUpResponse
                    {
                        Success = false,
                        Error = "Password and confirm password do not match",
                        ErrorCode = "S03",
                    };
                }
                var salt = PasswordHelper.GetSecureSalt();
                var passwordHash = PasswordHelper.HashUsingPbkdf2(userDto.Password, salt);
            user.Email = userDto.Email;
            user.FirstName = userDto.FirstName;
            user.LastName = userDto.LastName;
            user.Password = passwordHash; 
            user.PasswordSalt = Convert.ToBase64String(salt);
            user.Ts = DateTime.Now;

            _context.Users.Update(user);
            _context.SaveChanges();
            return new SignUpResponse {
                Success = true, 
                        Email = user.Email
            };
        }


        // Delete a user
        public  bool DeleteUser(int id)
        {
            var user = _context.Users.Find(id);

            if (user == null ) return false;
            var token = _context.RefreshTokens.FirstOrDefault(e => e.UserId == user.Id); 
        if (token != null) {

           _context.RefreshTokens.Remove(token);
        }
        _context.Users.Remove(user);
            
            _context.SaveChanges();
            return true;
        }
    }
}

File: frontend/src/Component/Service/UserService.js
import axios from "axios";
import { toast } from "sonner";
import config from "../config";

const API_BASE_URL = `${config.baseUrl}/User`;

const UserService = {
  getAllUsers: async () => {
    try {
      const response = await axios.get(`${API_BASE_URL}`);
      return response.data;
    } catch (error) {
      console.error("Error fetching users:", error);
      toast.error(error);
      throw error;
    }
  },

  getUserById: async (userId) => {
    try {
      if (userId <= 0) {
        throw new Error("Invalid user ID.");
      }

      const response = await axios.get(`${API_BASE_URL}/${userId}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching user with ID ${userId}:`, error);
      throw error;
    }
  },

  createUser: async (userData) => {
    try {
      if (!userData) {
        throw new Error("User data is required.");
      }

      const response = await axios.post(`${API_BASE_URL}`, userData);
      return response.data;
    } catch (error) {
      console.error("Error creating user:", error);
      throw error;
    }
  },

  updateUser: async (userId, userData) => {
    try {
      if (userId <= 0) {
        throw new Error("Invalid user ID.");
      }

      const response = await axios.put(`${API_BASE_URL}/${userId}`, userData);
      return response.data;
    } catch (error) {
      console.error(`Error updating user with ID ${userId}:`, error);
      throw error;
    }
  },
  updateCurrentUser: async (userId, userData) => {
    try {
      const response = await axios.put(
        `${API_BASE_URL}/currentUser/${userId}`,
        userData,
      );
      return response;
    } catch (error) {
      throw error;
    }
  },
  deleteUser: async (userId) => {
    try {
      if (userId <= 0) {
        throw new Error("Invalid user ID.");
      }

      await axios.delete(`${API_BASE_URL}/${userId}`);
      return true;
    } catch (error) {
      console.error(`Error deleting user with ID ${userId}:`, error);
      throw error;
    }
  },
};

export default UserService;
